### DISCUSSION TITLE
Which options / commands to downscale (resize to smaller) image with best possible quality?

### DISCUSSION THREAD

**Discussion by FurkanGozukara** [ImageMagick/ImageMagick#3852] [Help]
Edit. I have tested these commands and uploaded all results to the github including photoshop resize as well.

So what other options I can add my commands to test out?

I am using **Version: ImageMagick 7.1.0-0 Q16 x64 2021-06-12 **

Github repository : https://github.com/FurkanGozukara/image-test

tested commands for monster: https://github.com/FurkanGozukara/image-test/blob/master/size_test/all_tested_Commands.txt

test results are in here : https://github.com/FurkanGozukara/image-test/tree/master/size_test

tested commands for map: https://github.com/FurkanGozukara/image-test/blob/master/size_test_maps/all_tested_Commands.txt

test results are in here : https://github.com/FurkanGozukara/image-test/tree/master/size_test_maps

;============================

I would like to resize images to smaller dimension with best possible algorithm

I don't care about speed / memory / time

I want to obtain best possible outcome even better than Photoshop 2021 automatic downscale algorithm

So which options should I try? Can you give me several examples?

I would like to try combination of all so that I can find the best possible combination

Here an image from my web based game that I want to evaluate

https://www.monstermmorpg.com/Vesuverex-Monster-Dex-18

![800px](https://static.monstermmorpg.com/images/monsters-images-800-800/18-Vesuverex.png) 

Since its width is bigger than its height, I want to downscale to width 30 pixel and height automatically with keeping width / height ratio

These are the output of photoshop but I wonder If I can obtain better quality with ImageMagick

![30px](https://static.monstermmorpg.com/images/monsters-images-30-30/18-Vesuverex.png) 

![60px](https://static.monstermmorpg.com/images/monsters-images-60-60/18-Vesuverex.png) 

![120px](https://static.monstermmorpg.com/images/monsters-images-120-120/18-Vesuverex.png) 

![180px](https://static.monstermmorpg.com/images/monsters-images-180-180/18-Vesuverex.png) 

![240px](https://static.monstermmorpg.com/images/monsters-images-240-240/18-Vesuverex.png) 

![300px](https://static.monstermmorpg.com/images/monsters-images-300-300/18-Vesuverex.png) 

And here kind of different image example

https://www.monstermmorpg.com/Maps-Lily-Pond

Full size : 
![1024px full size](https://static.monstermmorpg.com/images/maps/Lily-Pond.png) 

![256px by photoshop](https://static.monstermmorpg.com/images/maps_256_px/Lily-Pond.png) 

![128px by photoshop](https://static.monstermmorpg.com/images/maps_128_px/Lily-Pond.png)

**Comment by fmw42** [comment]
Please fix your images. They do not show. Imagemagick will likely not make better images than Photoshop. There is little improvement available that would be visible to the naked eye. Also Photoshop typically adds some unsharp masking besides doing resampling. But you can read about different approaches to removing artifacts at https://legacy.imagemagick.org/Usage/filter/ and especially at https://legacy.imagemagick.org/Usage/filter/nicolas/ Approaches are often image dependent on the type of image you are processing.

**Reply by FurkanGozukara** [reply]
I can see all images in windows - chrome browser. There are some commands here https://legacy.imagemagick.org/Usage/filter/nicolas/ I may test all of them I guess

how can i set 1 dimension instead of percent?

like 30 pixel either height or width whichever one is bigger

**Reply by fmw42** [reply]
I cannot see any images you posted on my browser.  Did you leave off the suffix on the files? All uploaded images, need to have suffixes (.jpg etc) attached or they do not show properly.

`-resize 30x30`

See https://imagemagick.org/script/command-line-processing.php#geometry

**Reply by FurkanGozukara** [reply]
images are coded like this on the first post `![800px](https://static.monstermmorpg.com/images/monsters-images-800-800/18-Vesuverex.webp) `

**Reply by FurkanGozukara** [reply]
ok now this command works

now i need to figure out how can I test all options to compare myself :D 

`convert "C:/temp/org.webp" -colorspace RGB -filter Lanczos -define filter:blur=.9891028367558475 -resize 30x30 -colorspace sRGB "C:/temp/org_30_30.webp"`

**Reply by fmw42** [reply]
I do not believe that WEBP images are generally supported by this forum.

**Reply by fmw42** [reply]
I have looked at your post and there are no suffixes attached. That likely means that this forum (GIT) does not recognize that format.

**Reply by FurkanGozukara** [reply]
very interesting. I can see all images on my chrome browser. Here let me show 

![image](https://user-images.githubusercontent.com/19240467/122990276-71ae0e00-d3ac-11eb-8d09-aff2b15552ea.png)

by the way I will edit extensions for png now so that you can see too

**Reply by FurkanGozukara** [reply]
@fmw42 I have found all filtering. However what else I can try? here my base convert test

`cmd /C convert "{0}" -colorspace RGB -filter {1} -resize {2} "{3}"`

And here the output it generates

![image](https://user-images.githubusercontent.com/19240467/122990549-bf2a7b00-d3ac-11eb-98c1-baac693b1d41.png)


Here the list of filter i tested

```
Bartlett
Blackman
Bohman
Box
Catrom
Cosine
Cubic
Gaussian
Hamming
Hann
Hermite
Jinc
Kaiser
Lagrange
Lanczos
Lanczos2
Lanczos2Sharp
LanczosRadius
LanczosSharp
Mitchell
Parzen
Point
Quadratic
Robidoux
RobidouxSharp
Sinc
SincFast
Spline
CubicSpline
Triangle
Welch
```

**So my questions is what else I can add to my test? So that I can obtain the best outcome**

**Reply by fmw42** [reply]
Not all browsers can see WEBP images. My old Safari does not. I had to view on Chrome. Your images are graphics not natural images. ImageMagick has -magnify to enlarge graphical images, but no function to reduce them other than -sample. Try -sample 30x30 or -filter point -resize 30x30.

**Reply by FurkanGozukara** [reply]
@fmw42 I am going to update my first message in few minutes. I will demonstrate which options I have tried and what results I have obtained

**Comment by fmw42** [comment]
What is Photoshop automatic resampling?  Do you mean you just batch processed a bunch of images? What resampling method did you use in Photoshop?

**Reply by FurkanGozukara** [reply]
In photoshop I just use resize image to smaller size with automatic options @fmw42 . I set width or height

**Reply by fmw42** [reply]
There must be an interpolation method option?  What was it?  Where do  you find automatic resize? Is that new in PS 2021?  I only have PS 2019.

**Reply by fmw42** [reply]
Have you tried -adaptive-resize?

**Reply by fmw42** [reply]
Have you seen https://legacy.imagemagick.org/Usage/filter/#photoshop

**Reply by fmw42** [reply]
OK. In PS 2019, there is an resample option of automatic when doing size changes. But I have no idea what they do. Typically PS adds some unsharp masking. See IM -unsharp. You can add that to your command line after doing the the resize. Other than that, I have no idea what to suggest.

**Reply by fmw42** [reply]
One other thought. When downsizing by a large amount, one technique is to do it in successive steps. Resize by 2x. Then resize that result by 2x, etc until you get close. Then resize the final amount to your exact size in pixels. You  can try that with any -resize method or just -scale to average blocks of two by two pixels at a time.

**Reply by FurkanGozukara** [reply]
@fmw42 thank you very much for answers. I have updated top of my first post can you see my updated post? here i have uploaded everything to a github repository : https://github.com/FurkanGozukara/image-test

**Reply by fmw42** [reply]
What did you conclude as the best method for your images?  Can you actually see much difference?  My eyes have not been that critical to judge in the past.  Did you try my other suggestions?  Did you try different other options for -resize as per the page from Nicolas Robidoux?  Did you try a simple -sample?  Did you try my successive resize approach using any of the -resize filter options?

**Reply by FurkanGozukara** [reply]
I haven't tried sample yet. How can I try that? Currently testing "-adaptive-unsharp", "-unsharp", "-define filter:blur=0.75" options .

**Comment by fmw42** [comment]
I found this article about PS resizing and what they do. See https://photoshopcafe.com/resize-images-photoshop-keep-quality/. Apparently, automatic is bi-cubic with sharpening. So use either mitchel or lanczos or catmul-rom with -resize and add -unsharp.

**Reply by fmw42** [reply]
I do not normally work with graphical images. So you might wait to hear back from some of the IM users who do work with that type of image.

**Reply by FurkanGozukara** [reply]
@fmw42 this is my base command

so i just need to add unsharp?

cmd /C convert "{input}" -quality 100 -filter {e.g. lanczos } -resize {dimension} "{output}"

**Reply by fmw42** [reply]
Just after -resize 30x30 add -unsharp with the options you want to use (such as -unsharp 0x1). See https://imagemagick.org/script/command-line-options.php#unsharp

**Reply by fmw42** [reply]
just replace -resize 30x30 with -sample 30x30 (there are no -filter options for it)

**Reply by FurkanGozukara** [reply]
@fmw42 it looks like I can't use `-unsharp` with `-define filter:blur=0.75` ??

the below command giving error

`convert "path\a.png" -quality 100 -filter Kaiser -define filter:blur=0.75 -resize 30x30 -unsharp "path\a_Kaiser_withBlur_adaptiveUnSharp.png"`

**Comment by fmw42** [comment]
Try unsetting the filter. 

`convert "path\a.png" -quality 100 -filter Kaiser -define filter:blur=0.75 -resize 30x30 +filter -define filter:blur=1 -adaptive-unsharp "path\a_Kaiser_withBlur_adaptiveUnSharp.png"`

**Reply by FurkanGozukara** [reply]
so we cant combine filter and unsharp?

**Reply by FurkanGozukara** [reply]
by the way this command also gives error

**Reply by fmw42** [reply]
There is no -adaptive-unsharpen.  It is -adaptive-sharpen and it needs arguments. See https://imagemagick.org/script/command-line-options.php#adaptive-sharpen.  Alternately, just -unsharpen with arguments.

**Reply by FurkanGozukara** [reply]
ok i have solved errors. will update github repository soon. some massive quality improvement with sharpen options

**Comment by FurkanGozukara** [comment]
@fmw42 here a quick comparison

photoshop resize
![image](https://user-images.githubusercontent.com/19240467/123005996-b8f1ca00-d3bf-11eb-92c8-dfd3218a4622.png)

vs

a_Lanczos2Sharp_withBlur_UnSharp
![image](https://user-images.githubusercontent.com/19240467/123006006-bc855100-d3bf-11eb-9ae0-e8aaad94a793.png)

**Reply by FurkanGozukara** [reply]
@fmw42 i have updated the entire repository with new testing

huge improvement compared to photoshop : https://github.com/FurkanGozukara/image-test

**Comment by snibgo** [comment]
> I would like to resize images to smaller dimension with best possible algorithm

But what does "best" mean?

Resizing always involves tradeoffs. Increasing sharpness (or "acutance", which isn't quite the same thing) is good but this usually increases staircasing ("jaggies") which is bad, and may increase overshoots ("halos").

And what works well for one image (such as a photograph) may not work well for another (such as graphic art).

I don't have a simple answer. Sorry.

**Reply by FurkanGozukara** [reply]
I found that Sinc and Lanczos are best algorithms

Moreover if your image shading is complex Sinc works better

However when you apply Unsharp on simple images, edges becomes a little bit worse

I did a lot of testing and looks like i will use different combinations for different target sizes
